File: backtracking/subsets.py
# Q:78 - https://leetcode.com/problems/subsets/

from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        subset = []
        
        def dfs(i):
            if i >= len(nums):
                result.append(subset.copy())
                return

            subset.append(nums[i])
            dfs(i + 1)

            subset.pop()
            dfs(i + 1)

        dfs(0)

        return result


c = Solution()

nums1 = [1,2,3]
nums2 = [0]

print(c.subsets(nums1))
# print(c.subsets(nums2))

File: backtracking/subsets.js
// Q:78 - https://leetcode.com/problems/subsets/

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
  let result = [];

  const dfs = (index, numsList) => {
    if (index === nums.length) {
      result.push([...numsList]);
      return;
    }

    numsList.push(nums[index]);
    dfs(index + 1, numsList);

    numsList.pop();
    dfs(index + 1, numsList);
  };

  dfs(0, []); 

  return result;
};

console.log(subsets([1, 2, 3]));